packages/protocol/contracts/common/Accounts.sol

```diff
@@ -284,40 +352,196 @@ contract Accounts is
+   * @param role The role that has been authorized.
+   */
+  function removeSigner(address signer, string memory role) public {
+    address defaultSigner = getDefaultSigner(msg.sender, role);
+    if (defaultSigner == signer) {
+      removeDefaultSigner(role);
+    }
+
+    delete accounts[msg.sender].signerAuthorizations[role][signer];
+    emit SignerRemoved(msg.sender, signer, role);
+  }
+
+  /**
+   * @notice Removes the signer for a default role.
+   * @param role The role that has been authorized.
+   */
+  function removeDefaultSigner(string memory role) public {
+    Account storage account = accounts[msg.sender];
+
+    if (keccak256(abi.encodePacked(role)) == keccak256(abi.encodePacked(ValidatorSigner))) {
+      account.signers.validator = address(0);
+    } else if (
+      keccak256(abi.encodePacked(role)) == keccak256(abi.encodePacked(AttestationSigner))
+    ) {
+      account.signers.attestation = address(0);
+    } else if (keccak256(abi.encodePacked(role)) == keccak256(abi.encodePacked(VoteSigner))) {
+      account.signers.vote = address(0);
+    }
+
+    emit DefaultSignerRemoved(msg.sender, account.defaultSigners[role], role);
+    account.defaultSigners[role] = address(0);
+  }
+
   /**
    * @notice Removes the currently authorized vote signer for the account.
    * Note that the signers cannot be reauthorized after they have been removed.
    */
   function removeVoteSigner() public {
-    Account storage account = accounts[msg.sender];
-    emit VoteSignerRemoved(msg.sender, account.signers.vote);
-    account.signers.vote = address(0);
+    address signer = getDefaultSigner(msg.sender, VoteSigner);
+    removeSigner(signer, VoteSigner);
+    emit VoteSignerRemoved(msg.sender, signer);
   }
 
   /**
    * @notice Removes the currently authorized validator signer for the account
    * Note that the signers cannot be reauthorized after they have been removed.
    */
   function removeValidatorSigner() public {
-    Account storage account = accounts[msg.sender];
-    emit ValidatorSignerRemoved(msg.sender, account.signers.validator);
-    account.signers.validator = address(0);
+    address signer = getDefaultSigner(msg.sender, ValidatorSigner);
+    removeSigner(signer, ValidatorSigner);
+    emit ValidatorSignerRemoved(msg.sender, signer);
   }
 
   /**
    * @notice Removes the currently authorized attestation signer for the account
    * Note that the signers cannot be reauthorized after they have been removed.
    */
   function removeAttestationSigner() public {
-    Account storage account = accounts[msg.sender];
-    emit AttestationSignerRemoved(msg.sender, account.signers.attestation);
-    account.signers.attestation = address(0);
+    address signer = getDefaultSigner(msg.sender, AttestationSigner);
+    removeSigner(signer, AttestationSigner);
+    emit AttestationSignerRemoved(msg.sender, signer);
+  }
+
+  function defaultSignerToAccount(address signer, string memory role)
```

Comment:
Can't this be just a simple `signerToAccount`?

Reply:
This is just extracting the common logic for each old `voteSignerToAccount`, `validatorSignerToAccount` and `attestationSignerToAccount` where we:

- perform signer to account
- verify the signer is for vote, validator or attestation respectively
- verify the signer is not an account.

Here I'm just keeping backwards compatibility

Reply:
I found this naming confusing as it does not actually depend on whether the signer is marked as the default. I'd suggest a name change, and based on the fact that this is intended to consolidate internal logic, a visibility change as well.